README.md
A Python library for selecting an item from a multi-field data list in terminal.

# Introduction

PyPick allows you to specify a list of data entries (each entry has multiple fields and is represented by a dict) and optional meta data on how each fields shold be shown and interactived with. It displays these entries in terminal, one row for each etnry and one column for each field. It returns an entry's data when user selects it.

PyPick provides a general API. You can customize it easily for different scenarios, for example:

- Select a host to ssh into (see screenshot below)
- Select a frequently accessed directory
- Select a frequently used command

![docs/images/demo.png](https://github.com/rayx/pypick/raw/master/docs/images/demo.png)

# Install

As Python2 will be EOL'ed soon, PyPick supports only Python3. To install it, run:

    $ python3 -m pip install pypick

# A Quick-Start Guide

## A Simplest Example

Suppose you have two hosts and you'd like to use PyPick to generate a list:

    hosts = [{'name': 'server-5',
              'host': '10.64.4.5',
              'user': ['root', 'rayx'],
              'project': 'project X',
              'description': 'ubuntu 16.04, 192G RAM'},
             {'name': 'server-66',
              'host': '10.64.4.66',
              'user': ['root'],
              'project': 'project Y',
              'description': 'centos 7.4, 96G RAM'}]

The simplest way to do it is to pass the data and specify the fields you'd like to show:

    from pypick import Pick
    
	# NOTE: Add the above 'hosts' variable definition here.

    fields = ['name', 'host', 'user', 'description']
    p = Pick(fields)
    p.add_entries(hosts)
    result = p.run()
    print(result)

The code generates the following list. Each entry takes one row, and each field is displayed in one column.

![docs/images/simplest_example.png](https://github.com/rayx/pypick/raw/master/docs/images/simplest_example.png)

As you may notice, there is a minor issue with the above list. The last field of the first entry is not completely shown. This is because PyPick divides the available space for all columns equally by default. To address the issue, you can either increase the width of your terminal window, or adjust width of other fields in a programmatic way. We'll talk more about this latter.

You can press 'UP' and 'DOWN' (or VI style 'j' and 'k') to navigate through items in the list, press 'ENTER' (or 'SPACE') to select an entry, or press 'ESC' (or 'q') to quit without selecting anyting. 

If you selects an entry, the code returns its value, containing only the fields you specified. For example, if you select the first entry, the data returned is:

    {'description': 'ubuntu 16.04', 'name': 'server-5', 'host': '10.64.4.5', 'user': 'root'}

## Non-Displayed Fields

PyPick allows you to specify fields that are not shown but are returned. For example, the following code generates the same list as above but it returns an additional 'project' field when user selects an entry.

    from pypick import Pick
    
    # NOTE: Add the above 'hosts' variable definition here.
    
    fields = ["name", "host", "user", "description"]
    extra_fields = ["project"]
    p = Pick(fields, extra_fields=extra_fields)
    p.add_entries(data_list)
    result = p.run()
    print(result)
	
## Customizing Field Appearance and Behavior

In previous example, PyPick determines column width for you and uses default foreground/background colors to show them. In practice you often know roughly the size of most fields. In that case, you can use 'width' attribute to specify their column width.

PyPick displays all fields in 'dark blue' foreground color and default background color by default. However, sometimes not all fields are of the same importance. For those less important fields (e.g, a comment field), you may want to visually deemphasize them. That can be done by using 'style' attribute, which is string representing a specific foreground and background color combintation.

When you select an entry, PyPick returns all these fields shown in the list by default. However, sometimes you may want to show a field but don't need its value (e.g., a comment field). In that case, you can set the field's 'return' attribute to False.

In our example, 'description' is a less important field because it provides only additional information not required by ssh command. The following code sets its 'style' attribute so that the field is displayed in dimmed foreground color and sets its 'return' attribute so that the field's value is not returned. The code also sets the width of 'name', 'host' and 'user' fields explicitly so that 'description' field will get more space than in the previous example.

    from pypick import Pick
    
	# NOTE: Add the above 'hosts' variable definition here.

    fields = ["name", "host", "user", "description"]
    field_attrs = {'name': {"width": 20},
                   'host': {"width": 20},
                   'user': {"width": 12},
                   'description': {'return': False,
                                   'style': 'trivial'}}
    p = Pick(fields, field_attrs=field_attrs)
    p.add_entries(hosts)
    result = p.run()
    print(result)

It generates a list like the following. Note 'description' field has more space and its text is displayed in gray white.

![docs/images/field_attributes.png](https://github.com/rayx/pypick/raw/master/docs/images/field_attributes.png)

There is one thing that might puzzle you and I haven't talked about. It's the small triangle after the data of 'user' field. It's an indicator showing that the field have multiple values. In our example, the first entry's 'user' field is list containg more than one values and hence the indicator appears.

If a field value is a list containg more than one items, PyPick displays the first value by default. It also defines a shortcut key allowing user to iterate over other available values. That shortcut key is the first character of the field name by default. So if you navigate to first entry and press 'u', the text in the 'user' field will be 'rayx' instead. If you then press 'ENTER' to select the entry, the code returns:

    {'user': 'rayx', 'host': '10.64.4.5', 'name': 'server-5'}
	
Note 'description' field is not returned because we set this field's 'return' attribute to False.

You can specify a custom shortcut by setting a field's 'shortcut' attribute.

## Defining Entry Attributes

In the last section we talk about defining field attributes to customize field appearance and behavior. In this section we'll talk about entry attributes, which affect an entry's (and hence its columns) appearance.

Entries can have parent/child relationship. Child entry is indented relative to its parent when shown in list. PyPick allows user define the parent/child relationship in a simple approach. Each entry can have a '_level' attribute, which defaults to 0. If entry B follows entry A (note that entries are always in order) and B's '_level' is larger than that of A by 1, then B is A's child and is indented properly.

Some entries can have higher priority than the others. You indicate that to PyPick by setting their '_critical' attributes to True. PyPick highlights those entries with a different foreground color.

Let's change the host data a bit to demonstrate the two features:

    hosts = [{'name': 'server-5',
              'host': '10.64.4.5',
              'user': ['root', 'rayx'],
              'project': 'project X',
              'description': 'ubuntu 16.04, 192G RAM'},
             {'name': 'vm-176',
              'host': '192.168.122.176',
              'user': ['root', 'rayx'],
              'description': 'QEMU test: hotplugging',
              '_level': 1},
             {'name': 'nested-vm',
              'host': '192.168.18.3',
              'user': ['root', 'rayx'],
              'description': 'QEMU test: nested vm',
              '_level': 2},
             {'name': 'vm-37',
              'host': '192.168.122.37',
              'user': ['root', 'rayx'],
              'description': 'QEMU test: virtio-block',
              '_level': 1},
             {'name': 'server-66',
              'host': '10.64.4.66',
              'user': ['root'],
              'project': 'project Y',
              'description': 'centos 7.4, 96G RAM'}]

With the other portion of the code unchanged, the new host data generates a list like the following:

![docs/images/entry_attributes.png](https://github.com/rayx/pypick/raw/master/docs/images/entry_attributes.png)

## Organizing Data Entries into Groups

So far, we add data entries to list directly in all examples and get flat lists. PyPick allows you to organize data entries into groups and add those groups to list. You can mix these two approaches if you like.

To demostrate this feature, let's suppose we have two host lists:

    group1 = [{'name': 'server-5',
               'host': '10.64.4.5',
               'user': ['root', 'rayx'],
               'project': 'project X',
               'description': 'ubuntu 16.04, 192G RAM'},
              {'name': 'server-66',
               'host': '10.64.4.66',
               'user': ['root'],
               'project': 'project Y',
               'description': 'centos 7.4, 96G RAM'}]
    
    group2 = [{'name': 'arm-server',
               'host': '10.64.37.55',
               'user': ['rayx', 'root'],
               'description': 'Centriq 2400, 192G RAM'},
              {'name': 'x86-server',
               'host': '10.64.37.51',
               'user': ['rayx', 'root'],
               'description': 'Xeon Gold 5118, 192G RAM'},
              {'name': 'test-client',
               'host': '10.64.37.182',
               'user': ['rayx', 'root']}]

The code adds them to two groups, respectively:

    from pypick import Pick

    # NOTE: Add the above 'group1' and 'group2' variable definition here

    fields = ['name', 'host', 'user', 'description']
    field_attrs = {'name': {"width": 20},
               'host': {"width": 20},
               'user': {"width": 12},
               'description': {'return': False,
                               'style': 'trivial'}}
	
    p = Pick(fields, field_attrs = field_attrs)
    g1 = p.create_group('Aarch64 Servers')
    g1.add_entries(group1)
	g2 = p.create_group('Benchmarking')
	g2.add_entries(group2)
	result = p.run()
	print(result)

It generates a list like the following:

![docs/images/group.png](https://github.com/rayx/pypick/raw/master/docs/images/group.png)

A group can override the global field attributes. That means a group can have its own rules on which fields to show and how to show them. For simplicity's sake I'll not talk the details here. Please refer to the library's API reference.

## Defining Your Own Theme

PyPick defines a few built-in sytles for you to customize field foreground/background colors. The styles are:

- 'focused': For fields of entry being focused
- 'critical': For fields of entries which have '_critical' set
- 'normal': For regular fields
- 'trivial': For less important fields
- 'misc': For other non-data text (group name, dash line, etc.)

The following are their default values:

    | Style      | Foreground Color | Background Color |
    |------------+------------------+------------------|
    | 'focused'  | 'white'          | 'dark blue'      |
    | 'critical' | 'dark red'       | 'default'        |
    | 'normal'   | 'dark blue'      | 'default'        |
    | 'trivial'  | 'default'        | 'default'        |
    | 'misc'     | 'default'        | 'default'        |

As mentioned above, you can set a field's 'style' attribute to one of them to customize that field's appearance.

What's more, you can also customize these styles default values by calling set_theme() function.

For example 'trivial' style uses default foreground color by default, if you want to deemphasize it further, you can set its foregroud color to 'dark gray' for example. 

Note to Solarized color theme user:

Due to the way how Solarized 16-color palette is organized, the 'dark gray' color is actually located at 'light green' slot. So you should set 'trivial' style's foreground color to 'ligh green' instead.

The following code changes 'trivial' and 'misc' styles. Note the code is written for Solarized color theme. If you aren't using it, perhaps you should consider it :) or adjust the code for your terminal's palette.

    from pypick import Pick, set_theme
    
    group1 = [{'name': 'server-5',
               'host': '10.64.4.5',
               'user': ['root', 'rayx'],
               'project': 'project X',
               'description': 'ubuntu 16.04, 192G RAM'},
              {'name': 'server-66',
               'host': '10.64.4.66',
               'user': ['root'],
               'project': 'project Y',
               'description': 'centos 7.4, 96G RAM'}]
    
    fields = ['name', 'host', 'user', 'description']
    field_attrs = {'name': {"width": 20},
                   'host': {"width": 20},
                   'user': {"width": 12},
                   'description': {'return': False,
                                   'style': 'trivial'}}

    # NOTE: an empty string ('') is interpreted as 'default'
    theme = {'trivial': ('light green', ''), 'misc': ('brown', '')}
    set_theme(theme)
    
    p = Pick(fields, field_attrs = field_attrs)
    g1 = p.create_group('ARM Servers')
    g1.add_entries(group1)
    result = p.run()
    print(result)

It generates a list like the following:

![docs/images/theme.png](https://github.com/rayx/pypick/raw/master/docs/images/theme.png)

# API Reference

For a more complete description on the concepts and API reference, please install the package and run:

    $ python3 -m pydoc pypick


CHANGELOG

CHANGELOG
=========

0.2.1: 2018-04-26
-----------------
- Removed the setup.py in package dir, which was checked in by mistake.

0.2.0: 2018-04-20
-----------------
- Initial official (beta) release.


tests/test_ui.py
from pypick.ui import Column, Item, Group, List, EventLoop

def test_basic():
    name_column_attrs = {"width":20,
                         "shortcut": None,
                         "return": True,
                         "style": "normal"}
    name = ("name", "ＷＩＤＥ－ＣＨＡＲＡＣＴＥＲ", name_column_attrs)

    user_column_attrs = {"width": 20,
                         "shortcut": 'u',
                         "return": True,
                         "style": "normal"}
    user = ("user", ["root", "rayx"], user_column_attrs)

    host_column_attrs = {"width":20,
                         "shortcut": None,
                         "return": True,
                         "style": "normal"}
    host = ("host", "192.168.1.18", host_column_attrs)

    desc_column_attrs = {"width": None,
                         "shortcut": None,
                         "return": False,
                         "style": "trivial"}
    desc = ("desc", "this is a comment.", desc_column_attrs)

    fields = [name, user, host, desc]
    extra_fields = {"field_X": "X", "field_Y": "Y"}

    item1 = Item(fields, {}, {})
    item2 = Item(fields, {}, {"_level": 1, "_critical": True})
    item3 = Item(fields, extra_fields, {"_level": 2})
    group1 = Group([item1, item2, item3], name="group1")

    item4 = Item(fields, {}, {})
    item5 = Item(fields, extra_fields, {"_level": 1})
    group2 = Group([item4, item5], name="group2")

    list = List([group1, group2])
    result = EventLoop(list).run()
    print(result)


tests/test_pick.py
from pypick import Pick

def test_basic():
    data_list = [{"name": "server-5",
                  "host": "10.64.4.5",
                  "user": ["root", "rayx"],
                  "project": "project X",
                  "description": "ubuntu 16.04"},
                 {"name": "server-66",
                  "host": "10.64.4.66",
                  "user": ["root"],
                  "project": "project Y",
                  "description": "centos 7.4"}]

    fields = ["name", "host", "user", "description"]
    p = Pick(fields)
    g1 = p.create_group("My Hosts")
    g1.add_entries(data_list)
    result = p.run()
    print(result)


pypick/ui.py
# Copyright (C) 2019 by Huan Xiong. All Rights Reserved.
# Licensed under GPLv3 or later. See LICENSE file under top level directory.

import signal
import sys
import urwid
from wcwidth import wcswidth


class NoSpace(Exception):
    """Raised if there isn't enough space to display a column"""
    pass


class Theme:
    DEFAULT = {'focused': ('white', 'dark blue'),
               'critical': ('dark red', ''),
               'normal': ('dark blue', ''),
               'trivial': ('', ''),
               'misc': ('', '')}
    COMPANION_SUFFIX = '_non_text'

    def __init__(self, theme={}):
        self.theme = sanitize_input(theme, self.DEFAULT)
        self.create_companions()

    def create_companions(self):
        # A style applies to all characters in a column, including the
        # 'blank space' (they are not blank, but padded with blank space
        # characters). So, if a style contains 'underline' attribute,
        # spaces in the column are also underlined. That causes the UI
        # ugly. To address this issue, for each style containing
        # 'underline' attribute, a companion style is created for it,
        # which has 'underline' removed. These companion styles are used
        # for displaying non-text characters (e.g., space, indicator,
        # etc.) in the column.
        for k, v in self.theme.copy().items():
            fg, bg = v
            if 'underline' in fg:
                new_fg = ','.join([i for i in fg.split(',')
                                   if i.strip() != 'underline'])
                # Add a companion for this style
                self.theme[k + self.COMPANION_SUFFIX] = (new_fg, bg)

    def get_companion(self, style):
        companion = style + self.COMPANION_SUFFIX
        if companion in self.theme:
            return companion
        return style

    def get_palette(self):
        palette = []
        for name, value in self.theme.items():
            fg, bg = value
            palette.append((name, fg, bg))
        return palette


class Column:
    """
    A Column instance is reponsible for showing, updating, and
    returning value of an entry's field.

    Args:
         item (object): Item widget instance to which this column belongs.
         index (int): 0-based index of the column, increasing from left
             to right.
         name (str): field name
         value_candicates (str or list): field value. Note user
             may pass a list of values.
         column_attrs (dict): a dict containing this field's attributes.
    """
    ATTRS = {'width': None,
             'style': 'normal',
             'shortcut': None,
             'return': True}
    SPACE = ' '
    LIST_INDICATOR = '▾'
    LEVEL_INDICATOR = '⤷'

    def __init__(self, item, index, name, value_candidates, column_attrs):
        self.item = item
        self.index = index
        self.name = name
        self.value_candidates = value_candidates
        self.value_index = 0

        column_attrs = sanitize_input(column_attrs, self.ATTRS)
        self.width = column_attrs['width']
        self.style = column_attrs['style']
        self.shortcut = column_attrs['shortcut']
        self.will_return = column_attrs['return']

        # Calculate value and shortcut. Shortcut is set only when
        # 1) user sets it, _and_ 2) the field value is a list containing
        # more than one items.
        shortcut = self.shortcut
        self.shortcut = None
        if isinstance(self.value_candidates, list):
            if value_candidates:
                self.value = str(value_candidates[0])
                if len(value_candidates) > 1:
                    self.shortcut = shortcut
                    # Assign a default shortcut based on column name
                    # if user doesn't specify one.
                    if not self.shortcut:
                        self.shortcut = self.name[0]
            else:
                self.value = ''
        else:
            self.value = str(value_candidates)

    def chop_text(self, u, width):
        """
        Remove characters from the end of a unicode string until its
        display width is equal or less than the given width. It works
        with wide characters which has variable  display width.

        Args:
            u (str): a unicode string
            width (int): a width limit

        Returns:
            str: A substr that meets with the width limit.
        """
        while len(u) > 0 and wcswidth(u) > width:
            u = u[0:-1]
        return u

    def get_text_and_attrs(self, focus=False, width=0):
        """Return text to be displayed in this column.

        Args:
            focus (boolen): If the item to which the column belongs is
                being focused or not.
            width (int): The maximum width that is allowd.

        Returns:
            bytes: UTF-8 encoding of the unicode string to be displayed.
            list: A list of display attributes in run-length encoding.
        """
        text = []
        attrs = []

        def _create_text_and_attrs(snippet, style):
            text_width = sum([wcswidth(t) for t in text])
            if text_width + wcswidth(snippet) >= width:
                snippet = self.chop_text(snippet, width - text_width - 1)
                if snippet:
                    text.append(snippet)
                    attrs.append((style, len(snippet)))
                raise NoSpace()
            else:
                text.append(snippet)
                attrs.append((style, len(snippet)))

        # Determine the style (fg/bg colors) to render the column's text
        # 1) Get it from user input (or use default)
        style = self.style
        # 2) Columns of a critical data item should use 'critical' style,
        # unless the column is a trivial one.
        if self.item.is_critical() and style != 'trivial':
            style = 'critical'
        # 3) Columns of a focused data item should use 'focused' style.
        if focus:
            style = 'focused'
        # Theme is a global variable.
        companion_style = theme.get_companion(style)

        # Generate text and its run-length encoded attributes
        try:
            # 1) Add two leading spaces in the first colum of an item
            if not self.index:
                _create_text_and_attrs(self.SPACE * 2, companion_style)
            # 2) Indent the text of the first column of a child item
            if self.item.get_level() and not self.index:
                indent = self.SPACE * self.item.get_level() * 2 + \
                    self.LEVEL_INDICATOR + self.SPACE
                _create_text_and_attrs(indent, companion_style)
            # 3) Add the column's value
            _create_text_and_attrs(self.value, style)
            # 4) Append an indicator if it has multiple values
            if self.shortcut:
                _create_text_and_attrs(self.LIST_INDICATOR, companion_style)
        except NoSpace:
            pass
        finally:
            # 5) Add Padding
            size = width - sum([wcswidth(t) for t in text])
            if size:
                text.append(self.SPACE * size)
                attrs.append((companion_style, size))

        # Urwid TextCanvas accepts bytes, instead of unicode string
        text_b = bytearray()
        for index, snippet in enumerate(text):
            snippet_b = snippet.encode()
            text_b.extend(snippet_b)
            style, _ = attrs[index]
            attrs[index] = (style, len(snippet_b))
        return bytes(text_b), attrs

    def update_value(self):
        """Get the next value in value list and set column value to it.

           The function is called only if self.shortcut is not None.
        """
        self.value_index = (self.value_index + 1) % len(self.value_candidates)
        self.value = self.value_candidates[self.value_index]


class Item(urwid.Widget):
    """
    An Item instance is a urwid widget which shows a data entry in
    columns (one column for each field).

    Args:
        columns (list of tuple): A list of column data. Each item
            represents a column, in this format: (name, value, attrs).

                name (str): Column name.
                Value (str): Column value. Note it can be a list.
                attrs (dict): A dict containing the column's attribute
                    name/value pairs.
        hidden_columns (dict): A dict containing non-displayed field
            name/value pairs.
        item_attrs (dict): A dict containing the data entry's attributes.

    Raises:
        NoSpace: Raised if there isn't enough space to show columns.
    """
    _sizing = frozenset(['flow'])
    ATTRS = {'_level': 0,
             '_critical': False}

    def __init__(self, columns, hidden_columns, item_attrs):
        super(Item, self).__init__()
        self.hidden_columns = hidden_columns
        self.item_attrs = sanitize_input(item_attrs, self.ATTRS)
        self.columns = self.create_columns(columns, item_attrs)

    def is_critical(self):
        return self.item_attrs['_critical']

    def get_level(self):
        return self.item_attrs['_level']

    def create_columns(self, columns, item_attrs):
        columnobjs = []
        for index, column in enumerate(columns):
            name, value_candidates, column_attrs = column
            columnobjs.append(Column(self, index, name, value_candidates,
                                     column_attrs))
        return columnobjs

    def rows(self, size, focus=False):
        return 1

    def render(self, size, focus=False):
        (maxcol, ) = size
        self.set_columns_width(maxcol)
        text, attrs = self.get_text_and_attributes(focus=focus)
        return urwid.TextCanvas(text, attrs, maxcol=maxcol)

    def get_text_and_attributes(self, focus=False):
        itemtext_b = bytearray()
        attrs = []
        for c in self.columns:
            width = c.width if c.width else c._width
            column_text, column_attrs = c.get_text_and_attrs(focus=focus,
                                                             width=width)
            itemtext_b.extend(column_text)
            attrs.extend(column_attrs)
        return [bytes(itemtext_b)], [attrs]

    def set_columns_width(self, maxcol):
        used = sum([c.width for c in self.columns if c.width])
        if used > maxcol:
            raise NoSpace("There isn't enough space to show all columns")

        no_width = [c for c in self.columns if not c.width]
        for c in no_width:
            c._width = int((maxcol - used) / len(no_width))

    def selectable(self):
        return True

    def keypress(self, size, key):
        global result
        if key in ('enter', ' '):
            result = self.hidden_columns
            for c in self.columns:
                if c.will_return:
                    result[c.name] = c.value
            raise urwid.ExitMainLoop()
        else:
            c = self.get_column_by_shortcut(key)
            if c:
                c.update_value()
                self._invalidate()
                return
        return key

    def get_column_by_shortcut(self, key):
        for c in self.columns:
            if key == c.shortcut:
                return c
        return None


class Group(urwid.Pile):
    def __init__(self, items, name=None):
        blank = urwid.Text('')
        title = urwid.AttrMap(urwid.Text('[ %s ]' % name), 'misc')
        divider = urwid.AttrMap(urwid.Divider('-'), 'misc')
        if name:
            inner_widgets = [blank, title, divider] + items
        else:
            inner_widgets = [blank] + items
        super(Group, self).__init__(inner_widgets)

        # Add VIM-like 'j' and 'k' key behavior
        cmd_map = urwid.CommandMap().copy()
        cmd_map['j'] = 'cursor down'
        cmd_map['k'] = 'cursor up'
        self._command_map = cmd_map


class List (urwid.ListBox):
    def __init__(self, groups):
        it = urwid.SimpleListWalker(groups)
        super(List, self).__init__(it)

        # Add VIM-like 'j' and 'k' key behavior
        cmd_map = urwid.CommandMap().copy()
        cmd_map['j'] = 'cursor down'
        cmd_map['k'] = 'cursor up'
        self._command_map = cmd_map


class EventLoop(urwid.MainLoop):
    def __init__(self, widget):
        super(EventLoop, self).__init__(widget,
                                        palette=theme.get_palette(),
                                        unhandled_input=self.global_keypress)

    def run(self):
        try:
            super(EventLoop, self).run()
        except KeyboardInterrupt:
            pass
        return result # This is a global variable in this module.

    def global_keypress(self, key):
        if key in ('q', 'esc'):
            raise urwid.ExitMainLoop()


def sanitize_input(user_input, spec):
    # Check unknown keys
    for name in user_input.keys():
        if name not in spec.keys():
            raise ValueError('Unknown key: %s in %s' %
                             (name, user_input))

    # Add missing items with their default values
    d = user_input.copy()
    for name, value in spec.items():
        if name not in d:
            d[name] = value
    return d


def set_theme(new_theme):
    """Set a custom theme.

    Args:
        theme (dict): A dict containing text style name/value pairs.
            Text style value is a tuple containing fg/bg colors.

    For more details see Theme section in 'pydoc3 pypick' command output.
    """
    global theme
    theme = Theme(new_theme)


theme = Theme()
result = None


pypick/__init__.py
# Copyright (C) 2019 by Huan Xiong. All Rights Reserved.
# Licensed under GPLv3 or later. See LICENSE file under top level directory.

"""
Selecting an item from a list of multi-field data in terminal

The module defines the following public interfaces:

    Classes:
        Pick: The main API of this module. A Pick instance takes a
            list of multi-field data entries, and optional meta data
            on each field's appearance and behavior. It displays data
            entries in text based UI, one row for each data entry and
            one column for each field. It returns an entry's data when
            user selects it.
        Group: Data entries can be added to list directly, or be organized
            into groups. A group contains a list of entries and can
            define its own rules on which fields to show and how to show
            them.

    Exceptions:
        NoSpace: Raised if there isn't enough space to display a column

    Functions:
        set_theme(): Customize the theme of the UI.

1. Data Entries
---------------

A data entry has mutliple fields and is represented by a dictionary.
For example, a data entry for SSH login may be like the follwoing:

    {'name': 'server-5',
     'host': '10.64.4.5',
     'user': ['root', 'rayx'],
     'project': 'project X',
     'description': 'ubuntu 16.04'}

From Pick's perspective, a data entry's fields fall into these categories:

    - Fields that are shown in UI.
    - Fields that aren't shown in UI, but will be returned when the
      entry is selected.
    - Other Fields. These fields are ignored by Pick.

For fields shown in UI, pick allows user to define their UI appearance
and behavior. See field attributes in next section.

A data entry may also have some special fields, which don't contain
data but entry attributes, like if this is high priorty data entry, or
if this is a child entry of the one prior to it. Entry attributes apply
to all fields of this entry and affect their UI. See entry attributes
in next section.

Data entries can be organized into groups. A group can override global
fields attributes and define its own rules on which fields to show and
how to show them.

2. Field Attributes and Entry Attirbutes
----------------------------------------

Meta data include field attributes and entry attributes (attributes
that apply to a data entry and hence all its fields).

Field attributes describes UI attributes of a field. They are
specified when creating a Pick instance. The following field
attributes are supported:

  - width (type: int, default: None):
        Width of the column to show the field. If not specified, the
        column width is calculated automatically.
  - style (type: str, default: 'normal'):
        Style of the column. See more on this below.
  - shortcut (type: str representing key press, default: see below):
        This is only useful if the field data is a list, in which
        case the first value in the list is showed by default.
        Shortcut is a key press that allows user to iterate over
        available values in the list. If field value is a list
        containing more than one items and user doesn't specify a
        shortcut, Pick will generate one automatically by using the
        first character of the field's name.

        The following are examples of valid shortcut value: 'u', 'U',
        ' ' (space), 'ctrl u', 'f1', 'right', etc.

  - return (type: boolean, default: True):
        Whether to return the field's data when the entry is selected.

Entry attributes describes entry wide attributes that may afffect
its UI. The following entry attributes are supported:

  - _level (type: int, default: 0):
        This is used to define the parent/child relationship between
        a data entry and the one prior to it.
  - _critical (type: boolean, default: False):
        A critical data entry is highlighted in UI.

3. Theme
--------

Depending on their attributes and states, fields may be displayed in
different foreground and background color combinations. The
foreground/background color combination is called text style in Pick.
A text style has a name so it can be referenced.

Pick defines five text styles:

  - 'focused': Text style to display item being focused
  - 'cirtical': Text style to display items which have '_critical' set
  - 'normal': Default text style to display fields
  - 'trivial': Text style to display unimportant ('trivial') fields
  - 'default': Text style to display title (e.g., group name, etc.)

User can reference these text style names in fields attributes.

User can customize these text style values (that is, their foreground
and background colors) by calling set_theme(). Text style value is
a tuple of this format:

  (fg_color, bg_color)

Please refer to urwid document at the following link for valid fg/bg
color values:

http://urwid.org/manual/displayattributes.html#standard-foreground-colors

4. Usage Example
----------------

The following is a simple example. It shows two entries. For each
entry, its 'name', 'host', 'user', 'description' fields are shown.
When user selects an entry, its 'name', 'host', 'user', and 'project'
fields are returned.

    $ cat test_pick.py
    from pypick import Pick

    data_list = [{'name': 'server-5',
                  'host': '10.64.4.5',
                  'user': ['root', 'rayx'],
                  'project': 'project X',
                  'description': 'ubuntu 16.04'},
                 {'name': 'server-66',
                  'host': '10.64.4.66',
                  'user': ['root'],
                  'project': 'project Y',
                  'description': 'centos 7.4'}]

    fields = ['name', 'host', 'user', 'description']
    extra_fields = ['project']
    field_attrs = {'description': {'return': False}}

    p = Pick(fields, extra_fields, field_attrs)
    g1 = p.create_group('group1')
    g1.add_entries(data_list)
    result = p.run()
    print(result)

    $ python3 test_pick.py
    {'user': 'rayx', 'project': 'project X', 'name': 'server-5',
     'host': '10.64.4.5'}
"""

from .pick import Pick, Group
from .ui import set_theme, NoSpace

__all__ = ['Pick', 'Group', 'NoSpace', 'set_theme']
__version__ = '0.2.1'
__license__ = 'GPLv3+'
__author__ = 'Huan Xiong <huan.xiong@outlook.com>'
